manager/shard_strategy.go

```diff
@@ -0,0 +1,226 @@
+		}
+
+		var envType environmentType
+		switch shardConfig.Type {
+		case config.ProjectShardType:
+			envType = project
+		case config.DomainShardType:
+			envType = domain
+		}
+
+		return &EnvironmentShardStrategy{
+			EnableUncoveredReplica: shardConfig.EnableUncoveredReplica,
+			EnvType:                envType,
+			Replicas:               replicas,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("shard strategy '%s' does not exist", shardConfig.Type)
+}
+
+// HashShardStrategy evenly assigns disjoint keyspace responsibilities over a collection of pods.
+// All FlyteWorkflows are assigned a shard-key using a hash of their executionID and are then
+// processed by the FlytePropeller instance responsible for that keyspace range.
+type HashShardStrategy struct {
+	EnableUncoveredReplica bool
+	PodCount               int
+}
+
+func (h *HashShardStrategy) GetPodCount() int {
+	if h.EnableUncoveredReplica {
+		return h.PodCount + 1
+	}
+
+	return h.PodCount
+}
+
+func (h *HashShardStrategy) HashCode() (uint32, error) {
+	return computeHashCode(config.HashShardType, h)
+}
+
+func (h *HashShardStrategy) UpdatePodSpec(pod *v1.PodSpec, containerName string, podIndex int) error {
+	container, err := getContainer(pod, containerName)
+	if err != nil {
+		return err
+	}
+
+	if podIndex >= 0 && podIndex < h.PodCount {
+		startKey, endKey := ComputeKeyRange(v1alpha1.ShardKeyspaceSize, h.PodCount, podIndex)
+		for i := startKey; i < endKey; i++ {
+			container.Args = append(container.Args, "--propeller.include-shard-key-label", fmt.Sprintf("%d", i))
+		}
+	} else if h.EnableUncoveredReplica && podIndex == h.PodCount {
+		for i := 0; i < v1alpha1.ShardKeyspaceSize; i++ {
+			container.Args = append(container.Args, "--propeller.exclude-shard-key-label", fmt.Sprintf("%d", i))
+		}
+	} else {
+		return fmt.Errorf("invalid podIndex '%d' out of range [0,%d)", podIndex, h.GetPodCount())
+	}
+
+	return nil
+}
+
+// Computes a [startKey, endKey) pair denoting the key responsibilities for the provided pod index
+// given the keyspaceSize and podCount parameters
+func ComputeKeyRange(keyspaceSize, podCount, podIndex int) (int, int) {
+	keysPerPod := keyspaceSize / podCount
+	keyRemainder := keyspaceSize - (podCount * keysPerPod)
+
+	return computeStartKey(keysPerPod, keyRemainder, podIndex), computeStartKey(keysPerPod, keyRemainder, podIndex+1)
+}
+
+func computeStartKey(keysPerPod, keysRemainder, podIndex int) int {
+	return (intMin(podIndex, keysRemainder) * (keysPerPod + 1)) + (intMax(0, podIndex-keysRemainder) * keysPerPod)
+}
+
+func intMin(a, b int) int {
+	if a < b {
+		return a
+	}
+
+	return b
+}
+
+func intMax(a, b int) int {
+	if a > b {
+		return a
+	}
+
+	return b
+}
+
+// The ProjectShardStrategy assigns project(s) to individual FlytePropeller instances to
+// determine FlyteWorkflow processing responsibility.
+type EnvironmentShardStrategy struct {
+	EnvType                environmentType
+	EnableUncoveredReplica bool
```

Comment:
how does this really work? i thought we are assigning bins to specific propellers right?

Reply:
resolved per our discussion